### PR TITLE
TASK Binary expression ast with only two operands: left and right

### PR DESCRIPTION
**Pull request by mhsdesign**
... instead of multiple operands

i inlined $left and $right into the BinaryOperationNode:
```php
final class BinaryOperationNode
{
    private function __construct(
        public readonly ExpressionNode $left,
        public readonly BinaryOperator $operator,
        public readonly ExpressionNode $right
    ) {
    }
}
```

this will make working with the BinaryOperandNode (eg type resolving/narrowing) much easier.
Also the php and js AST also use the fields $left $right and $operator instead of a collection (even when the same operator is used).

as a sideeffect we now also transpile `true === true === true` correctly to php (by wrapping it in parenthesis): `(true === true) === true`
without parenthesis it is not valid php (its valid ecmascript though ^^)

The above mentioned case with `=== true` is obviously an edge case and currently `1 + 1 + 1` is also transpiled with parenthesis: `(1 + 1) + 1` which are optional and not needed by php - we still need to decide if we like this see https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/16#discussion_r1174619323.

### CODE REVIEW

File: test/Integration/Examples/Numbers/Numbers.php [hunk 1]
```diff
@@ -10,6 +10,6 @@ final class Numbers extends BaseClass
 {
     public function render(): string
     {
-        return (string) (0 + 1234567890 + 42 + 0b10000000000000000000000000000000 + 0b01111111100000000000000000000000 + 0b00000000011111111111111111111111 + 0o755 + 0o644 + 0xFFFFFFFFFFFFFFFFF + 0x123456789ABCDEF + 0xA + 1E3 + 2e6 + 123.456 + 0.1e2 + .22);
+        return (string) (((((((((((((((0 + 1234567890) + 42) + 0b10000000000000000000000000000000) + 0b01111111100000000000000000000000) + 0b00000000011111111111111111111111) + 0o755) + 0o644) + 0xFFFFFFFFFFFFFFFFF) + 0x123456789ABCDEF) + 0xA) + 1E3) + 2e6) + 123.456) + 0.1e2) + .22);
```
Comment (mhsdesign) [inline test/Integration/Examples/Numbers/Numbers.php:13]: if you prefer not having those parenthesis here, i experimented with a `shouldAddParenthesisIfNecessary` flag, this way we could transpile without parenthesis (if the operand is the same in chained binary ops)

```php
public function transpile(BinaryOperationNode $binaryOperationNode): string
{
    $expressionTranspiler = new ExpressionTranspiler(
        scope: $this->scope,
        shouldAddQuotesIfNecessary: true,
        shouldAddParenthesisIfNecessary: !(
            $binaryOperationNode->operands->first->root instanceof BinaryOperationNode
            && $binaryOperationNode->operands->first->root->operator === $binaryOperationNode->operator
        )
    );

    $first = $expressionTranspiler->transpile($binaryOperationNode->operands->first);
    $operator = $this->transpileBinaryOperator($binaryOperationNode->operator);
    $second = $expressionTranspiler->transpile($binaryOperationNode->operands->second);

    if ($this->shouldAddParenthesis) {
        return sprintf('(%s %s %s)', $first, $operator, $second);
    } else {
        return sprintf('%s %s %s', $first, $operator, $second);
    }
}
```

**Comment by mhsdesign** [comment]
This pr will also make implementing https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/8 easier

File: test/Integration/Examples/Numbers/Numbers.php (hunk 1 above)
Comment (grebaldi) [reply test/Integration/Examples/Numbers/Numbers.php:13]: I don't mind those :) All's good :+1:

**Review by grebaldi — APPROVED** [review]
- Tests are green
- One fewer class
- Waaaay better, semantically
- Everything gets easier

What's  not to like? :smiley: 

Thanks a lot @mhsdesign!

**Comment by mhsdesign** [comment]
:D

**Pull request merged**